src/BSeries.jl

```diff
@@ -1453,4 +1456,379 @@ include("latexify.jl")
     include("precompile.jl")
 end
 
+
+
+"""
+        is_energy_preserving(rka::RungeKuttaMethod,order)::Bool
+
+This function checks whether a Runge-Kutta method `rk` is 
+energy-preserving for a given `order`.
+"""
+function is_energy_preserving(rk::RungeKuttaMethod,order)
+#generate bseries 
+    series = bseries(rk, order)
+    #pass it to the main function 'is_energy_preserving'
+    is_energy_preserving(series)
+end
+
+"""
+    is_energy_preserving(map_series)::Bool
+This function checks whether the B-series `series` of a time integration
+method is energy-preserving for Hamiltonian systems - up to the
+[`order`](@ref) of the `series`.
+
+# References
+This code is based on the Theorem 2 of
+- Elena Celledoni, Robert I. McLachlan, Brynjulf Owren, and G. R. W. Quispel. 
+  "Energy-preserving integrators and the structure of B-series."
+  Foundations of Computational Mathematics 10 (2010): 673-693.
+  [DOI: 10.1007/s10208-010-9073-1](https://link.springer.com/article/10.1007/s10208-010-9073-1)
+ """
+function is_energy_preserving(map_series)
+    flow_series = modified_equation(map_series)
+    #check if the first element is map or flow.
+    #renormalize by multiplying by symmetry factor
+    renormalize_bseries!(flow_series)
+    #save all the coefficients in an array: it is easier
+    #to call an array full of fractions when working with the 
+    #functions 'renormalize_bseries!' and 'energy_preserving_trees_test' 
+    coefficients = collect(values(flow_series))
+    #save all the RootedTrees in another array: 
+    thetrees = collect(keys(flow_series))
+    #we need only the level sequence
+    # Create an empty vector to store the converted trees into arrays
+    trees = Vector{Vector{Int}}(undef, length(flow_series))
+    # Convert the trees and store them in the 'trees' vector
+    for i in 1:length(flow_series)
+        trees[i] = thetrees[i].level_sequence
+    end
+    energy_preserving_flag = true
+    max_length = length(trees[end])
+    l = length(trees)
+    #first, check if the energy_preserving conditon is satisfied
+    #low order (provided that the computational cost at order < 5)
+    # is low. This is because the 'rank' function's behavior
+    # in the test function doesn't works by separate for low orders.
+    if low_order_energy_preserving(trees, coefficients) == false
+        return false
+    end
+    #loop over same-order trees
+    order = 5
+    while order <= max_length && energy_preserving_flag == true
+        same_order_trees = []
+        same_order_coeffs = []
+        #we create and fill the arrays of trees and coefficients
+        #corresponding to an 'order'
+        for j in 1:l
```

Comment:
Instead of this, why not just use `RootedTreeIterator(order)`?

Then also you can get the coefficients via `flow_series[t]` where `t` is a tree.  So then I think there is no need to make the separate arrays of trees and coefficients.

Reply:
Sorry, I don't see how we can use `RootedTreeIterator(order)` here

Reply:
You need all the trees of a specific order, right?  `RootedTreeIterator(order)` gives you exactly that.

Reply:
```
same_order_trees = collect(RootedTreesIterator(order))
same_order_coeffs = [flow_series[tree] for tree in same_order_trees]
```

Reply:
@ranocha My code above doesn't work, and I don't quite understand why.  Can you tell us what is the right way to get an array from the iterator?

<img width="372" alt="Screen Shot 2023-06-13 at 11 43 47 AM" src="https://github.com/ranocha/BSeries.jl/assets/697763/6a3f665e-a4b6-4f19-b012-141ff2d2ffa1">

Reply:
The tree `t` is a view to an internal buffer for efficiency. You can use something like
```julia
julia> map(copy, RootedTreeIterator(4))
4-element Vector{RootedTree{Int64, Vector{Int64}}}:
 RootedTree{Int64}: [1, 2, 3, 4]
 RootedTree{Int64}: [1, 2, 3, 3]
 RootedTree{Int64}: [1, 2, 3, 2]
 RootedTree{Int64}: [1, 2, 2, 2]
```

See also https://github.com/SciML/RootedTrees.jl#iteration-over-rootedtrees and https://sciml.github.io/RootedTrees.jl/stable/api_reference/#RootedTrees.RootedTreeIterator

Reply:
Thanks!